File: intro-to-akka-demo/src/main/java/be/yannickdeturck/akkademo/example/Example3Supervisor.java
package be.yannickdeturck.akkademo.example;

import akka.actor.*;
import akka.japi.pf.DeciderBuilder;
import akka.japi.pf.ReceiveBuilder;
import scala.concurrent.duration.Duration;

import java.util.Random;

/**
 * Example of a Supervisor Strategy
 */
public class Example3Supervisor {
    public static void main(String[] args) {
        ActorSystem system = ActorSystem.create();
        final ActorRef supervisor = system.actorOf(Supervisor.props(), "supervisor");
        for (int i=0; i < 10; i++) {
            supervisor.tell(new UnstableActor.Command(), ActorRef.noSender());
        }
        system.awaitTermination();
    }
}

class Supervisor extends AbstractLoggingActor {
    public static final OneForOneStrategy STRATEGY = new OneForOneStrategy( // each child is treated separately
            10, // max 10 restart per 10 seconds
            Duration.create("10 seconds"),
            DeciderBuilder
                    .match(RuntimeException.class, ex -> SupervisorStrategy.restart())
                    .build()
    );


    public Supervisor() {
        final ActorRef unstableActor = getContext().actorOf(UnstableActor.props(), "unstableActor");
        receive(ReceiveBuilder
                .matchAny(any -> unstableActor.forward(any, getContext()))
                .build()
        );
    }

    @Override
    public SupervisorStrategy supervisorStrategy() {
        return STRATEGY;
    }

    public static Props props() {
        return Props.create(Supervisor.class);
    }
}

class UnstableActor extends AbstractLoggingActor {

    public static class Command {}

    public UnstableActor() {
        log().info("Creating UnstableActor...");
        receive(ReceiveBuilder
                .match(Command.class, this::onCommand)
                .build()
        );
    }

    private void onCommand(Command c) {
        int rng = new Random().nextInt(10);
        if (rng < 9) {
            throw new RuntimeException("Woops, something went wrong");
        }
        log().info("Handling the command...");
    }

    public static Props props() {
        return Props.create(UnstableActor.class);
    }

    @Override
    public void postStop() throws Exception {
        log().info("Stopping UnstableActor...");
    }
}

File: intro-to-akka-demo-scala/src/main/scala/be/yannickdeturck/akkademo/example/Example3Supervisor.scala
package be.yannickdeturck.akkademo.example

import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, OneForOneStrategy, Props, SupervisorStrategy}

import scala.concurrent.duration._
import scala.util.Random

object Example3Supervisor extends App {
  val system = ActorSystem("system")
  val supervisor = system.actorOf(Supervisor.props, "supervisor")
  for (i <- 0 until 10){
    supervisor ! UnstableActor.Command
  }
}

class Supervisor extends Actor with ActorLogging {
  val unstableActor: ActorRef = context.actorOf(UnstableActor.props, "unstableActor")

  override def supervisorStrategy: SupervisorStrategy = {
    OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 second) {
      case _: RuntimeException         => SupervisorStrategy.Restart
    }
  }

  override def receive: Receive = {
    case msg => unstableActor forward msg
  }
}

object Supervisor {
  val props: Props = Props[Supervisor]
}

class UnstableActor extends Actor with ActorLogging {
  override def receive: Receive = {
    case UnstableActor.Command =>
      val rng = Random.nextInt(10)
      if (rng < 4) throw new RuntimeException("Woops, something went wrong")
      log.info("Handling the command...")
  }

  @scala.throws[Exception](classOf[Exception])
  override def postStop(): Unit = log.info("Stopping UnstableActor...")

  @scala.throws[Exception](classOf[Exception])
  override def preStart(): Unit = log.info("Creating UnstableActor...")
}

object UnstableActor {
  case class Command()
  val props: Props = Props[UnstableActor]
}